server/routes/warningDetails.ts

```diff
@@ -1,10 +1,287 @@
+        requirementTypeSubCategoryDescription: enforceableContactWithRequirement.requirement.subType.description,
+        rejectionReason: req.body[bodyParamBreachReason],
+        fromDate: null,
+        toDate: null,
+      }
+
+      requirementList.push(applyContactDateRangesToRequirement(currentRequirement, warningDetails.enforceableContacts))
+    })
+
+    breachNotice.breachNoticeRequirementList = requirementList
+    const errorMessages: ErrorMessages = validateWarningDetails(breachNotice, req.body.responseRequiredByDate)
+
+    const hasErrors: boolean = Object.keys(errorMessages).length > 0
+
+    // if we dont have validation errors navigate to ...next screen
+    if (!hasErrors) {
+      breachNotice.warningDetailsSaved = true
+      await breachNoticeApiClient.updateBreachNotice(id, breachNotice)
+      if (req.body.action === 'saveProgressAndClose') {
+        res.send(`<script nonce="${res.locals.cspNonce}">window.close()</script>`)
+      } else if (req.body.action === 'refreshFromNdelius') {
+        // redirect to warning details to force a reload
+        res.redirect(`/warning-details/${id}`)
+      } else {
+        res.redirect(`/next-appointment/${id}`)
+      }
+    } else {
+      const failuresRecorded = createSelectItemListFromEnforceableContacts(warningDetails.enforceableContacts)
+      const breachReasons = convertReferenceDataListToSelectItemList(warningDetails.breachReasons)
+      const requirementsList = createFailuresBeingEnforcedRequirementSelectList(
+        warningDetails.enforceableContacts,
+        warningDetails.breachReasons,
+        breachNotice,
+      )
+
+      res.render(`pages/warning-details`, {
+        breachNotice,
+        warningDetails,
+        failuresRecorded,
+        breachReasons,
+        requirementsList,
+        currentPage,
+        errorMessages,
+      })
+    }
+  })
+
+  function applyContactDateRangesToRequirement(
+    breachNoticeRequirement: BreachNoticeRequirement,
+    enforceableContactList: EnforceableContact[],
+  ): BreachNoticeRequirement {
+    const dateList = enforceableContactList
+      .filter(i => i.requirement.id.toString() === breachNoticeRequirement.requirementId.toString())
+      .map(i => i.datetime)
+    const maxDate = dateList.reduce((a, b) => (a > b ? a : b))
+    const minDate = dateList.reduce((a, b) => (a < b ? a : b))
+    return {
+      ...breachNoticeRequirement,
+      fromDate: minDate.toString(),
+      toDate: maxDate.toString(),
+    }
+  }
+
+  function validateWarningDetails(breachNotice: BreachNotice, responseRequiredByDate: string): ErrorMessages {
+    const errorMessages: ErrorMessages = {}
+    const currentDateAtStartOfTheDay: LocalDateTime = LocalDate.now().atStartOfDay()
+    try {
+      // eslint-disable-next-line no-param-reassign
+      breachNotice.responseRequiredDate = fromUserDate(responseRequiredByDate)
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    } catch (error: unknown) {
```

Comment:
Maybe log the error here, in case we need to figure out the reason for it later?  Then you don't need the eslint ignore line too.
```suggestion
    } catch (error: unknown) {
      logger.error(error)
```